mllib/src/main/scala/org/apache/spark/mllib/pmml/export/TreeModelUtils.scala

```diff
@@ -0,0 +1,275 @@
+      .withScore(mlLibNode.predict.predict.toString)
+
+    val predicate = getPredicate(mlLibNode)
+    if (predicate.isDefined) {
+      node.withPredicate(predicate.get)
+    }
+
+    node
+  }
+
+  /** Get PMML mining function type for given mlLib Algo. */
+  private def getPMMLMiningFunctionType(mlLibAlgo: Algo): MiningFunctionType = {
+    mlLibAlgo match {
+      case Algo.Classification => MiningFunctionType.CLASSIFICATION
+      case Algo.Regression => MiningFunctionType.REGRESSION
+    }
+  }
+
+  /** Get PMML datafield based on the mllib split feature. */
+  private def getDataField(mllibNode: Node): Option[DataField] = {
+    if (!mllibNode.isLeaf && mllibNode.split.isDefined) {
+      val split = mllibNode.split.get
+      val dataField = new DataField()
+        .withName(FieldName.create("field_" + split.feature.toString))
+        .withDataType(DataType.fromValue(split.threshold.getClass.getSimpleName.toLowerCase))
+        .withOpType(OpType.fromValue(split.featureType.toString.toLowerCase))
+
+      split.featureType match {
+        case FeatureType.Continuous => dataField.withOpType(OpType.CONTINUOUS)
+        case FeatureType.Categorical => {
+          dataField.withOpType(OpType.CATEGORICAL)
+          val categories = split.categories
+            .map(category => new org.dmg.pmml.Value(category.toString)).asJava
+          dataField.withValues(categories)
+        }
+      }
+
+      Some(dataField)
+    }
+    else {
+      None
+    }
+  }
+
+  /** Get PMML Mining field based on Mllib node split feature. */
+  private def getMiningField(mllibNode: Node): Option[MiningField] = {
+    if (!mllibNode.isLeaf && mllibNode.split.isDefined) {
+      val split = mllibNode.split.get
+
+      val miningField = new MiningField()
+        .withName(FieldName.create("field_" + split.feature.toString))
+        .withUsageType(FieldUsageType.ACTIVE)
+
+      Some(miningField)
+    }
+    else {
+      None
+    }
+
+  }
+
+  /** Get distinct PMML mining fields list for a given mlLib decision tree model. */
+  def getMiningFieldsForTree(treeModel: DecisionTreeModel): List[MiningField] = {
+
+    @tailrec
+    def getMiningFieldsForTreeRec(
+                                   nodeList: List[Node],
+                                   miningFlds: List[MiningField]): List[MiningField] = {
+
+      nodeList match {
+        case Nil => miningFlds
+        case nd :: ls if (nd.isLeaf) => miningFlds
+        case nd :: ls if (!nd.isLeaf && nd.split.isDefined) => {
+          val ndList = MutableList[Node]()
+          if (nd.leftNode.isDefined) {
+            ndList += nd.leftNode.get
+          }
+
+          if (nd.rightNode.isDefined) {
+            ndList += nd.rightNode.get
+          }
+
+          getMiningFieldsForTreeRec(ndList.toList, miningFlds :+ getMiningField(nd).get)
+
+        }
+      }
+    }
+
+    val miningFields = getMiningFieldsForTreeRec(List(treeModel.topNode), List[MiningField]())
```

Comment:
getMiningFieldsForTreeRec is very similar to getDataFieldsForTreeRec, couldn't you merge these?
Also, I am not sure you even need getMiningFieldsForTreeRec, you could create the mining fields from the data fields you already have, usually (in the models I exported) the field names were the same for mining fields and data fields.